src/base/memory_monitor_sr.cpp

```diff
@@ -69,28 +102,71 @@ namespace cubmem
     return allocated_size;
   }
 
-  std::string memory_monitor::make_tag_name (const char *file, const int line)
+  void memory_monitor::add_stat (char *ptr, const size_t size, const char *file, const int line)
+  {
+    std::string tag_name;
+    MMON_METAINFO metainfo;
+
+    // size should not be 0 because of MMON_ALLOC_META_SIZE
+    assert (size > 0);
+
+    metainfo.allocated_size = (uint64_t) size;
+    m_total_mem_usage += metainfo.allocated_size;
+
+    tag_name = make_tag_name (file, line);
+
+    std::unique_lock <std::mutex> tag_map_lock (m_tag_map_mutex);
+    auto search = m_tag_map.find (tag_name);
+    if (search != m_tag_map.end ())
+      {
+	metainfo.tag_id = search->second;
+	m_stat_map[metainfo.tag_id] += metainfo.allocated_size;
+      }
+    else
+      {
+	metainfo.tag_id = m_tag_map.size ();
+	// tag_id starts with 0
+	m_tag_map.emplace (tag_name, metainfo.tag_id);
+	m_stat_map.emplace (metainfo.tag_id, metainfo.allocated_size);
+      }
+    tag_map_lock.unlock ();
+
+    // put meta info into the allocated chunk
+    char *meta_ptr = get_meta_ptr (ptr, metainfo.allocated_size);
+    metainfo.magic_number = m_magic_number;
+    memcpy (meta_ptr, &metainfo, MMON_ALLOC_META_SIZE);
+    m_meta_alloc_count++;
+  }
+
+  void memory_monitor::sub_stat (char *ptr)
   {
-    std::string filecopy (file);
 #if defined(WINDOWS)
-    std::string target ("\\src\\");
+    size_t allocated_size = 0;
+    assert (false);
 #else
-    std::string target ("/src/");
-#endif // WINDOWS
+    size_t allocated_size = malloc_usable_size ((void *)ptr);
+#endif // !WINDOWS
 
-    // Find the last occurrence of "src" in the path
-    size_t pos = filecopy.rfind (target);
-#if !defined (NDEBUG)
-    size_t lpos = filecopy.find (target);
-    assert (pos == lpos);
-#endif // NDEBUG
+    assert (ptr != NULL);
 
-    if (pos != std::string::npos)
+    if (allocated_size >= MMON_ALLOC_META_SIZE)
```

Comment:
```suggestion
    if (allocated_size > MMON_ALLOC_META_SIZE)
```

Reply:
NIT:

To remove nested if statement.

```
if (allocated_size > MMON_ALLOC_META_SIZE)
{
    return;
}

...
```

Reply:
Even when malloc is requested to allocate memory of size 0, it returns a valid pointer instead of nullptr. Additionally, it is possible to free this allocated memory. Therefore, the situation where `allocated_size == MMON_METAINFO_SIZE` should also be considered.